File: packages/ogdf.js/src/module/src/EdgeInsertion.cpp
#include "../main.h"

#include <ogdf/planarity/FixedEmbeddingInserter.h>
#include <ogdf/planarity/MultiEdgeApproxInserter.h>
#include <ogdf/planarity/VariableEmbeddingInserter.h>

EM_PORT_API(FixedEmbeddingInserter *)
EdgeInsertionModule_FixedEmbeddingInserter(bool keepEmbedding, double percentMostCrossed, int removeReinsert, double timeLimit)
{
    FixedEmbeddingInserter *fixed = new FixedEmbeddingInserter();
    fixed->keepEmbedding(keepEmbedding);
    fixed->percentMostCrossed(percentMostCrossed);
    fixed->removeReinsert(static_cast<RemoveReinsertType>(removeReinsert));
    fixed->timeLimit(timeLimit);
    return fixed;
}

EM_PORT_API(MultiEdgeApproxInserter *)
EdgeInsertionModule_MultiEdgeApproxInserter(double percentMostCrossedFix, double percentMostCrossedVar, int removeReinsertFix, int removeReinsertVar, bool statistics, double timeLimit)
{
    MultiEdgeApproxInserter *multi = new MultiEdgeApproxInserter();
    multi->percentMostCrossedFix(percentMostCrossedFix);
    multi->percentMostCrossedVar(percentMostCrossedVar);
    multi->removeReinsertFix(static_cast<RemoveReinsertType>(removeReinsertFix));
    multi->removeReinsertVar(static_cast<RemoveReinsertType>(removeReinsertVar));
    multi->statistics(statistics);
    multi->timeLimit(timeLimit);
    return multi;
}

EM_PORT_API(VariableEmbeddingInserter *)
EdgeInsertionModule_VariableEmbeddingInserter(double percentMostCrossed, int removeReinsert, double timeLimit)
{
    VariableEmbeddingInserter *var = new VariableEmbeddingInserter();
    var->percentMostCrossed(percentMostCrossed);
    var->removeReinsert(static_cast<RemoveReinsertType>(removeReinsert));
    var->timeLimit(timeLimit);
    return var;
}

File: packages/ogdf.js/src/module/config/EdgeInsertion.js
import createModule from "../module-helper";
import { PARAMETER_TYPE } from "../../utils/parameter-type";

const NAME = "EdgeInsertionModule"

const MODULE_DIRECTORY = {
    FixedEmbeddingInserter: {
        keepEmbedding: {
            type: PARAMETER_TYPE.BOOL,
            range: [true, false],
            default: false
        },
        percentMostCrossed: {
            type: PARAMETER_TYPE.DOUBLE,
            range: [-Infinity, Infinity],
            default: 25
        },
        removeReinsert: {
            type: PARAMETER_TYPE.CATEGORICAL,
            range: ['None', 'Inserted', 'MostCrossed', 'All', 'Incremental', 'IncInserted'],
            default: 'None'
        },
        timeLimit: {
            type: PARAMETER_TYPE.DOUBLE,
            range: [-Infinity, Infinity],
            default: 1000
            // default: -1 if timeLimit < 0 will be unstoppable
        }
    },
    MultiEdgeApproxInserter: {
        percentMostCrossedFix: {
            type: PARAMETER_TYPE.DOUBLE,
            range: [-Infinity, Infinity],
            default: 25
        },
        percentMostCrossedVar: {
            type: PARAMETER_TYPE.DOUBLE,
            range: [-Infinity, Infinity],
            default: 25
        },
        removeReinsertFix: {
            type: PARAMETER_TYPE.CATEGORICAL,
            range: ['None', 'Inserted', 'MostCrossed', 'All', 'Incremental', 'IncInserted'],
            default: 'None'
        },
        removeReinsertVar: {
            type: PARAMETER_TYPE.CATEGORICAL,
            range: ['None', 'Inserted', 'MostCrossed', 'All', 'Incremental', 'IncInserted'],
            default: 'None'
        },
        statistics: {
            type: PARAMETER_TYPE.BOOL,
            range: [true, false],
            default: false
        },
        timeLimit: {
            type: PARAMETER_TYPE.DOUBLE,
            range: [-Infinity, Infinity],
            default: 1000
            // default: -1 if timeLimit < 0 will be unstoppable
        }
    },
    VariableEmbeddingInserter: {
        percentMostCrossed: {
            type: PARAMETER_TYPE.DOUBLE,
            range: [-Infinity, Infinity],
            default: 25
        },
        removeReinsert: {
            type: PARAMETER_TYPE.CATEGORICAL,
            range: ['None', 'Inserted', 'MostCrossed', 'All', 'Incremental', 'IncInserted'],
            default: 'None'
        },
        timeLimit: {
            type: PARAMETER_TYPE.DOUBLE,
            range: [-Infinity, Infinity],
            default: 1000
            // default: -1 if timeLimit < 0 will be unstoppable
        }
    }
}

const EdgeInsertionModule = createModule(NAME, MODULE_DIRECTORY)
export default EdgeInsertionModule